lightning-block-sync/src/convert.rs

```diff
@@ -0,0 +1,472 @@
+#[derive(Deserialize)]
+struct GetHeaderResponse {
+	pub version: i32,
+	pub merkleroot: String,
+	pub time: u32,
+	pub nonce: u32,
+	pub bits: String,
+	pub previousblockhash: String,
+
+	pub chainwork: String,
+	pub height: u32,
+}
+
+/// Converts from `GetHeaderResponse` to `BlockHeaderData`.
+impl TryFrom<GetHeaderResponse> for BlockHeaderData {
+	type Error = bitcoin::hashes::hex::Error;
+
+	fn try_from(response: GetHeaderResponse) -> Result<Self, bitcoin::hashes::hex::Error> {
+		Ok(BlockHeaderData {
+			header: BlockHeader {
+				version: response.version,
+				prev_blockhash: BlockHash::from_hex(&response.previousblockhash)?,
+				merkle_root: TxMerkleNode::from_hex(&response.merkleroot)?,
+				time: response.time,
+				bits: u32::from_be_bytes(<[u8; 4]>::from_hex(&response.bits)?),
+				nonce: response.nonce,
+			},
+			chainwork: hex_to_uint256(&response.chainwork)?,
+			height: response.height,
+		})
+	}
+}
+
+
+/// Converts a JSON value into a block. Assumes the block is hex-encoded in a JSON string.
+impl TryInto<Block> for JsonResponse {
+	type Error = std::io::Error;
+
+	fn try_into(self) -> std::io::Result<Block> {
+		match self.0.as_str() {
+			None => Err(std::io::Error::new(std::io::ErrorKind::InvalidData, "expected JSON string")),
+			Some(hex_data) => match Vec::<u8>::from_hex(hex_data) {
+				Err(_) => Err(std::io::Error::new(std::io::ErrorKind::InvalidData, "invalid hex data")),
+				Ok(block_data) => match encode::deserialize(&block_data) {
+					Err(_) => Err(std::io::Error::new(std::io::ErrorKind::InvalidData, "invalid block data")),
+					Ok(block) => Ok(block),
+				},
+			},
+		}
+	}
+}
+
+/// Converts a JSON value into the best block hash and optional height.
+impl TryInto<(BlockHash, Option<u32>)> for JsonResponse {
+	type Error = std::io::Error;
+
+	fn try_into(self) -> std::io::Result<(BlockHash, Option<u32>)> {
+		if !self.0.is_object() {
+			return Err(std::io::Error::new(std::io::ErrorKind::InvalidData, "expected JSON object"));
+		}
+
+		let hash = match &self.0["bestblockhash"] {
+			serde_json::Value::String(hex_data) => match BlockHash::from_hex(&hex_data) {
+				Err(_) => return Err(std::io::Error::new(std::io::ErrorKind::InvalidData, "invalid hex data")),
+				Ok(block_hash) => block_hash,
+			},
+			_ => return Err(std::io::Error::new(std::io::ErrorKind::InvalidData, "expected JSON string")),
+		};
+
+		let height = match &self.0["blocks"] {
+			serde_json::Value::Null => None,
+			serde_json::Value::Number(height) => match height.as_u64() {
+				None => return Err(std::io::Error::new(std::io::ErrorKind::InvalidData, "invalid height")),
+				Some(height) => match height.try_into() {
+					Err(_) => return Err(std::io::Error::new(std::io::ErrorKind::InvalidData, "invalid height")),
+					Ok(height) => Some(height),
+				}
+			},
+			_ => return Err(std::io::Error::new(std::io::ErrorKind::InvalidData, "expected JSON number")),
+		};
+
+		Ok((hash, height))
+	}
+}
+
+#[cfg(test)]
```

Comment:
Can you extract some test-cases from a real bitcoin node response to the various queries used?

Reply:
That's essentially what I'm doing here, primarily with the genesis block and generated programmatically (e.g. line 170) to ensure the tests are concise. I could add missing fields that are irrelevant to the conversion, but I don't think that adds much value to the tests.

Reply:
Right, I meant using the full set of queries used, eg including `GetHeaderResponse`.

Reply:
Hmm... not sure I understand. I believe that I covered all the queries. Did I miss something?

get_header:
- into_block_header_from_json_response_with_valid_header
- into_block_header_from_json_response_with_valid_header_array
- into_block_header_from_json_response_without_previous_block_hash

get_block
- into_block_from_valid_binary_response
- into_block_from_json_response_with_valid_block_data

get_best_block
- into_block_hash_from_json_response_without_height
- into_block_hash_from_json_response_with_height

Reply:
Right, I was looking for a string copied from the wire off wireshark or the like, but I suppose creating them by filling in objects manually is fine enough.

Reply:
For unit tests, I prefer to keep the inputs as concise as possible yet still representative. Using raw wire data would be reasonable for an integration test, IMHO.